src/Microsoft.Azure.CosmosRepository/Specification/Builder/SpecificationBuilder.cs

```diff
@@ -0,0 +1,134 @@
+using System.Linq.Expressions;
+using System.Text;
+using Microsoft.Azure.CosmosRepository.Paging;
+
+namespace Microsoft.Azure.CosmosRepository.Specification
+{
+    class SpecificationBuilder<T,TResult> : ISpecificationBuilder<T,TResult>
+        where T : IItem
+        where TResult : IQueryResult<T>
+
+    {
+        public BaseSpecification<T,TResult> Specification { get; }
+        public SpecificationBuilder(BaseSpecification<T, TResult> specification)
+        {
+            Specification = specification;
+        }
+    }
+    /// <summary>
+    /// 
+    /// </summary>
+    public static class SpecificationBuilderExtensions
+    {
+        /// <summary>
+        /// Add a search filter added to the query. Multiple filters will be evaluated togheter
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="specificationBuilder"></param>
+        /// <param name="criteria"></param>
+        public static ISpecificationBuilder<T, TResult> Where<T, TResult>(
+            this ISpecificationBuilder<T, TResult> specificationBuilder,
+            Expression<Func<T, bool>> criteria)
+            where T : IItem
+            where TResult : IQueryResult<T>
+
+        {
+            ((List<WhereExpressionInfo<T>>)specificationBuilder.Specification.WhereExpressions).Add(new WhereExpressionInfo<T>(criteria));
+
+            return specificationBuilder;
+        }
+
+        /// <summary>
+        /// Specify the query result will be ordered by <paramref name="orderExpression"/> in an ascending order
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="specificationBuilder"></param>
+        /// <param name="orderExpression"></param>
+        public static IOrderedSpecificationBuilder<T, TResult> OrderBy<T, TResult>(
+            this ISpecificationBuilder<T, TResult> specificationBuilder,
+            Expression<Func<T, object>> orderExpression)
+            where T : IItem
+            where TResult : IQueryResult<T>
+        {
+            ((List<OrderExpressionInfo<T>>)specificationBuilder.Specification.OrderExpressions).Add(new OrderExpressionInfo<T>(orderExpression, OrderTypeEnum.OrderBy));
+
+            OrderedSpecificationBuilder<T, TResult> orderedSpecificationBuilder = new(specificationBuilder.Specification);
+
+            return orderedSpecificationBuilder;
+        }
+
+        /// <summary>
+        /// Specify the query result will be ordered by <paramref name="orderExpression"/> in a descending order
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="specificationBuilder"></param>
+        /// <param name="orderExpression"></param>
+        public static IOrderedSpecificationBuilder<T, TResult> OrderByDescending<T, TResult>(
```

Comment:
This seems to imply you could do something such as `Query.Orderby(x => ...)` should you be able to do this before a `.Where` clause?

Reply:
Yes, these should be able to come in any order right? In Linq I could specify the order before there where clause